### PR TITLE
nixos/photoprism: init module

### PR DESCRIPTION
**Pull request by Stunkymonkey**
###### Description of changes

having module-options for `photoprism`
currently it is missing the mysql setup, but i suggest adding this later and default to `sqlite`.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Stunkymonkey** [comment]
@ofborg test photoprism

File: nixos/modules/services/web-apps/photoprism.nix
Comment (AndersonTorres) [inline nixos/modules/services/web-apps/photoprism.nix:29]:
```suggestion
    enable = mkEnableOption (lib.mdDoc "Photoprism web server");
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]
Otherwise LGTM

**Comment by nyanloutre** [comment]
Tested it on my system successfully

**Comment by Stunkymonkey** [comment]
@nyanloutre you suggest renaming `extraConfig` to `settings`? because this does not expect a string.

File: nixos/modules/services/web-apps/photoprism.nix
Comment (Stunkymonkey) [reply nixos/modules/services/web-apps/photoprism.nix:27]: not sure why that is a bad idea. I am against this suggestion. This will result in much more code and everything uses `lib`...

File: nixos/modules/services/web-apps/photoprism.nix
Comment (Stunkymonkey) [reply nixos/modules/services/web-apps/photoprism.nix:106]: I copied it from the papaerless module. Did not think about it. I can remove it as well...

File: nixos/modules/services/web-apps/photoprism.nix
Comment (Stunkymonkey) [reply nixos/modules/services/web-apps/photoprism.nix:27]: I agree that in combination with `let` this is not clear.
But here the scope is very clear.
If you look at packages and search for e.g. "meta = with lib; {"
```
$ grep -r "meta = with lib; {" | wc -l
18937
```
So this kind is used nearly everywhere.

Comment (Stunkymonkey) [reply nixos/modules/services/web-apps/photoprism.nix]: should be fixed now.

**Review by pennae — COMMENTED** [review]
looks good. always nice to see service hardening! :)

just one little nit left.

File: nixos/modules/services/web-apps/photoprism.nix
Comment (Stunkymonkey) [reply nixos/modules/services/web-apps/photoprism.nix:133]: the problem here is when photoprism is started the first time.
The admin account gets created when the first photoprism command is executed. So when executing the migration this already takes place.
So this will result in
```
preStart = ''
        ln -sf ${manage} photoprism-manage

        ${lib.optionalString (cfg.passwordFile != null) ''
          export PHOTOPRISM_ADMIN_PASSWORD=$(cat "$CREDENTIALS_DIRECTORY/PHOTOPRISM_ADMIN_PASSWORD")
        ''}
        exec ${cfg.package}/bin/photoprism migrations run -f
      '';

      script = ''
        ${lib.optionalString (cfg.passwordFile != null) ''
          export PHOTOPRISM_ADMIN_PASSWORD=$(cat "$CREDENTIALS_DIRECTORY/PHOTOPRISM_ADMIN_PASSWORD")
        ''}
        exec ${cfg.package}/bin/photoprism start
      '';
```
What do you think about this? Maybe its nice to have it correct in the first place and if the fixes arrive simple removals make everything nice. Otherwise we might forget about it.

Comment (Stunkymonkey) [reply nixos/modules/services/web-apps/photoprism.nix]: Is solved.

**Review by pennae — APPROVED** [review]

**Review by pennae — CHANGES_REQUESTED** [review]

File: nixos/modules/services/web-apps/photoprism.nix
Comment (Stunkymonkey) [reply nixos/modules/services/web-apps/photoprism.nix]: i did not test this, but added your suggestion :+1:

**Review by AndersonTorres — CHANGES_REQUESTED** [review]
About the commit sequence:

- a commit to add the package
- a commit to add module(s)
- a commit to add test(s)

**Comment by AndersonTorres** [comment]
> @nyanloutre you suggest renaming `extraConfig` to `settings`? I only tested the manage command with `passwd` and it worked without the `nsenter`.

Always when possible, structured `settings` should be preferred against stringly `extraConfig`.

**Comment by Stunkymonkey** [comment]
> > @nyanloutre you suggest renaming `extraConfig` to `settings`? I only tested the manage command with `passwd` and it worked without the `nsenter`.
> 
> Always when possible, structured `settings` should be preferred against stringly `extraConfig`.

the extraConfig here is not a string. But it gets appended to the minimal config. Therefore I thought it would be better to name it `extra`. But sure i renamed it to settings.

If you tell me everything is as wanted, then I am going to split the commits into separate ones.

**Review by AndersonTorres — APPROVED** [review]
Aesthetically, lgtm.

**Review by pennae — APPROVED** [review]

File: nixos/doc/manual/release-notes/rl-2305.section.md
Comment (AndersonTorres) [inline nixos/doc/manual/release-notes/rl-2305.section.md:37]:
```suggestion
- [photoprism](https://photoprism.app/), a AI-Powered Photos App for the Decentralized Web. Available as [services.photoprism](options.html#opt-services.photoprism.enable).
```

The phrase makes it look as if "we" = "Nixpkgs team".

**Review by AndersonTorres — CHANGES_REQUESTED** [review]
Otherwise lgtm

File: nixos/modules/services/web-apps/photoprism.nix
Comment (RaitoBezarius) [inline nixos/modules/services/web-apps/photoprism.nix:86]: getting-started

**Review by RaitoBezarius — APPROVED** [review]
Awesome, modulo some nit-picks!

**Comment by Stunkymonkey** [comment]
I am waiting for @AndersonTorres approval and then merge it. I guess three people are more then enough for a first version.

**Comment by RaitoBezarius** [comment]
> I am waiting for @AndersonTorres approval and then merge it. I guess three people are more then enough for a first version.

I think he already said LGTM modulo some nitpicks, once CI passes, I will merge it.

**Review by AndersonTorres — APPROVED** [review]

File: nixos/modules/services/web-apps/photoprism.nix
Comment (ajs124) [inline nixos/modules/services/web-apps/photoprism.nix:55]: was this intended to be a mandatory settings? IMO it shouldn't be, but right now the eval fails if it's not set, because the type isn't `nullOr`.

Comment (Stunkymonkey) [reply nixos/modules/services/web-apps/photoprism.nix:55]: photoprism can not start without this parameter... Therefore i made it mandatory.

**Comment by azrdev** [comment]
> Maybe add a wrapper script around nsenter to manage the database easily. I had to to this manually to reindex faces:
> 
> ```
> sudo nsenter -t 467850 -m -S 64600 -G 64600 --wdns=/var/lib/photoprism /var/lib/photoprism/photoprism-manage faces index
> ```

That comment is so useful! Is there a proper way to build a wrapper script which does this to call photoprism  management commands from any user (with the appropriate privileges)? I'm thinking like `nextcloud-occ`

**Pull request merged**
